### DISCUSSION TITLE
Strings as arguments

### DISCUSSION THREAD

**Discussion by saidctb** [pyccel/pyccel#1800] [Ideas]
Many functions in Blas and Lapack rely on passing strings as arguments, this is not currently handled in Pyccel e.g.
```python
def f(st:'str'):
    print(st)

g = epyccel(f)
/home/said/phd/pyccel/__epyccel__/__pyccel__/bind_c_mod_9w6hwod9oltk.f90:11:24:

   11 |   subroutine bind_c_f(st) bind(c)
      |                        1
Error: Character argument ‘st’ at (1) must be length 1 because procedure ‘bind_c_f’ is BIND(C)
```
It would be nice if we could handle passing a string or a single character.

**Comment by EmilyBourne** [comment]
Adding single characters would be useful for BLAS/LaPack, but I'm not sure how you would annotate this :thinking: there is no `char` type in Python

**Reply by saidctb** [reply]
yes I think we need  `char` and `char[:]`

**Comment by EmilyBourne** [comment]
When we do put string support back I'm not sure that we will do this using the built-in string support in Fortran because it is full of pitfalls (like the one you show above). Does this mean that we need a separate type for external libraries?
But maybe this type will also be needed for the bind c file

**Comment by saidctb** [comment]
In my opinion, we can have 3 separate types,  a single character `char` an array of characters  `char[:]` and the `str` type, the first two have their equivalents in C/Fortran, and the third one we will define ourselves in the targeted languages.

**Reply by EmilyBourne** [reply]
That makes sense to me and it lines up very well with the recent changes to the typing (there is already an internal `CharType`)

**Reply by EmilyBourne** [reply]
Do you know which of these objects are needed for lapack? All of them? Or just `char`?

**Reply by saidctb** [reply]
`char` is needed for sure, the others I'm not sure, probably not.

**Reply by EmilyBourne** [reply]
Ok. I think that has a fairly obvious type annotation too so I think we can go ahead and open a dedicated issue for that step. This will be useful for @jalalium

**Reply by EmilyBourne** [reply]
#1801

**Comment by saidctb** [comment]
We also have to handle array of chars see this example in Lapack https://github.com/Reference-LAPACK/lapack/blob/122506cd8b6ce050a200920c3d4c0b153b150fd8/SRC/xerbla.f#L79
so probably the user can create them using this
```python
string = np.empty((2, 5), dtype='c')
string[0] = "ABCDE"
string[1] = "FGHIJ"
print("string[0]=",string[0].tobytes())
```

**Reply by EmilyBourne** [reply]
Is this valid Python code? I didn't realise NumPy had support for characters. Although it seems logical given how easy it is to add.
This is also something that we could support relatively easily

**Reply by saidctb** [reply]
Yes, it is valid Python code, you can even specify the length of the strings by doing something like
```python
string = np.empty((2, 3), dtype='U5')
```
which is an array of strings of length 5.
this we can also support, but it can wait

**Reply by saidctb** [reply]
The simple case of a multidimensional array of one character should be easy to implement along with the issue https://github.com/pyccel/pyccel/issues/1801, and its type annotation should be `char[:,:,...]`

**Reply by saidctb** [reply]
For an array of fixed-sized strings we can have something like `U5[:,:]`
